rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,641 @@
+       applyMigrationState(result);
+       unsubscribe();
+     });
+   }
+}
+```
+
+Notes:
+  - To be sure that we do not encounter a race condition with the event subscribing / sending (workers subscribing after 
+    the main worker actually sent the `migration-complete` event and then waiting indefinitely), we are using the `persist`
+    option of the `broadcast` API. We felt this was a better approach than the alternative of having shared state among workers.
+
+## 5.5 Sharing state between workers
+
+This is not identified as necessary at the moment, and IPC broadcast should be sufficient, hopefully. We prefer to avoid
+the added complexity and risk of implicit dependencies if possible.
+
+If we do eventually need shared state, we would probably have to use syscall libraries to share buffers such as
+[mmap-io](https://www.npmjs.com/package/mmap-io), and expose a higher level API for that from the `clustering` service. More
+research would be required if this proved to be a necessity.
+
+# 6. Technical impact
+
+This section attempts to be an exhaustive inventory of the changes that would be required to support clustering mode.
+
+## 6.1 Technical impact on Core
+
+### 6.1.1 Handling multi-process logs
+
+This is an example of log output in a 2 workers cluster, coming from the POC:
+
+```
+[2021-03-02T10:23:41.834+01:00][INFO ][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.840+01:00][INFO ][plugins-service] Plugin initialization disabled.
+[2021-03-02T10:23:41.900+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
+[2021-03-02T10:23:41.903+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
+```
+
+The workers logs are interleaved, and, most importantly, there is no way to see which process each log entry is coming from. 
+We will need to address that.
+
+#### Options we considered:
+
+1. Having a distinct logging configuration (with separate log files) for each worker
+2. Centralizing log collection in the coordinator and writing all logs to a single file (or stdout)
+
+#### Our recommended approach:
+
+Overall we recommend keeping a single log file (option 2), and centralizing the logging system in the coordinator,
+with each worker sending the coordinator log messages via IPC. While this is a more complex implementation in terms
+of our logging system, it solves several problems:
+- Preserves backwards compatibility.
+- Avoids the issue of interleaved log messages that could occur with multiple processes writing to the same file or stdout.
+- Provides a solution for the rolling-file appender (see below), as the coordinator would handle rolling all log file
+(small caveat being a slight loss in file size precision).
```

Comment:
If only the coordinator is writing to disk, is this still true?

Reply:
Oops, you're right. Got some threads crossed here, removed.